---
title: "Tutorial I - Navigating the genetic dataset"
author: "Fan Wang"
date: "2025-05-31"
output: html_document
---

##### This tutorial introduces the structure, content, and usage of PLINK binary genotype datasets with 4,000 individuals and more than 300 K SNPs. The goal is to help beginners understand how genotype data is stored and accessed in PLINK, and prepare them for downstream tasks like GWAS and QC.

You will need to install [PLINK](https://zzz.bwh.harvard.edu/plink/) and run the analyses. PLINK is a free, open-source whole genome association analysis toolset, designed to perform a range of basic, large-scale analyses in a computationally efficient manner. For this tutorial, we are using simulated case/control data.

### 1. What is a PLINK Binary Dataset?

A PLINK binary dataset consists of three files:

- **gwas.qc.bed** - Binary genotype matrix (compressed, encoded format) 
- **gwas.qc.bim** - SNP annotation file: chromosome, SNP ID, base-pair position, alleles 
- **gwas.qc.fam** - Sample information file: IDs, sex, phenotype

```{r setup, include=FALSE}
# Set the working directory for the entire R Markdown file
knitr::opts_knit$set(root.dir = "/Users/macbook/work/Code/Dataset2/")
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE, warning=FALSE, results='hide'}
library(data.table)
library(ggplot2)
library(seqminer)
library(HardyWeinberg)
library(dplyr)
```
### 2. View the .fam and .bim Files

```{bash}
# Sample information file: IDs, sex, phenotype
cd /Users/macbook/work/Code/Dataset2/
head gwa.qc.fam
```

| **Column Name** | **Description** |
|----------------|--------------------------------------------------------|
| **FID** | Family ID (set to 0 if not used) |
| **IID** | Individual ID (unique for each person) |
| **PAT** | Paternal ID (set to 0 if unknown) |
| **MAT** | Maternal ID (set to 0 if unknown) |
| **SEX** | Sex: `1 = male`, `2 = female`, `0 = unknown` |
| **PHENOTYPE** | Trait status: `1 = control`, `2 = case`, `0` or `-9` = missing , or a quantitative trait |

```{bash}
# SNP annotation file: chromosome, SNP ID, base-pair position, alleles 
head gwa.qc.bim
```

| **Column Name** | **Description**                 |
|-----------------|---------------------------------|
| **CHR**         | Chromosome number               |
| **SNP**         | SNP ID (e.g., rs number)        |
| **CM**          | Genetic distance (usually 0)    |
| **BP**          | Physical position in base pairs |
| **A1**          | PLINK’s effect allele (the counted allele)|
| **A2**          | the other allele (reference/alternative allele depending on context)      |

### 3. Understanding the .bed file

The .bed file stores encoded genotypes in a compact binary format. It is not human-readable. Use PLINK to decode it.

Convert to readable format:

```{bash,results='hide',message=FALSE, warning=FALSE}
./plink2 --bfile gwa.qc --recode A --out gwa_qc_additive
```

This creates a .raw file with dosage coding:

0 = homozygous for A2

1 = heterozygous

2 = homozygous for A1

```{r}
# Load genotype data
geno <- fread("gwa_qc_additive.raw")  # columns include FID, IID, and SNP genotype
geno[1:5,1:10]
```
---

### 4. Hardy-Weinberg Equilibrium (HWE) test

We can test each SNP for deviation from Hardy-Weinberg Equilibrium:

```{bash,results='hide',message=FALSE, warning=FALSE}

./plink2 --bfile gwa.qc --hardy --out gwa_qc_hwe

```

This produces gwa_qc_hwe.hardy containing observed and expected genotype counts plus p-values.
```{r,message=FALSE, warning=FALSE}
# View HWE results
hwe <- fread("gwa_qc_hwe.hardy")
head(hwe)
```

CHROM – Chromosome number.

ID – SNP ID (e.g. rsID).

A1, AX – The two alleles (effect and reference).

HOM_A1_CT – Count of A1/A1 homozygotes.

HET_A1_CT – Count of A1/AX heterozygotes.

TWO_AX_CT – Count of AX/AX homozygotes.

O(HET_A1) – Observed heterozygote frequency (proportion).

E(HET_A1) – Expected heterozygote frequency under HWE.

P – Hardy-Weinberg p-value testing deviation from equilibrium.

For QC, SNPs with very low P are often removed, as this may indicate genotyping error.


### 5. Estimate Allele Frequencies
Calculate and output allele frequencies in your dataset:
```{bash,results='hide',message=FALSE, warning=FALSE}
./plink2 --bfile gwa.qc --freq --out gwa_qc_freq
```
View frequency table: gwa_qc_freq.afreq includes columns for SNP, chromosome, allele counts, MAF.
```{r,message=FALSE, warning=FALSE}
freq <- fread("gwa_qc_freq.afreq")
head(freq)
```

### 6. Subset SNPs and Individuals

#### Step 1: Create SNP and sample list files

```{bash,message=FALSE, warning=FALSE}
> snplist.txt
# Create SNP list file with 3 example SNPs
echo rs1525873 >> snplist.txt
echo rs2902663 >> snplist.txt
echo rs343376 >> snplist.txt
echo rs3813199 >> snplist.txt

# Create sample list file with header + 3 samples
#echo "FID IID" > samplist.txt
#echo "0 A2001" >> samplist.txt
#echo "2 A2003" >> samplist.txt
#echo "5 A2006" >> samplist.txt

```

#### Step 2: Subset SNPs and/or samples

```{bash, results='hide',message=FALSE, warning=FALSE}
./plink2 --bfile gwa.qc \
      --extract snplist.txt \
      --recode A \
      --out geno.additive
# --keep samplist.txt can be used to extract a specific list of samples
```

```{r}
# Load genotype data
geno_subset <- fread("geno.additive.raw")
geno_subset=na.omit(geno_subset)
geno_subset$PHENOTYPE=geno_subset$PHENOTYPE-1
head(geno_subset)
table(geno_subset$PHENOTYPE)
```


##### Alternatively, we can load all three files into R. 
The PLINK2 commands are especially useful when working with a large number of SNPs (genome-wide), while R is better suited for loading smaller subsets of genotype data.
```{r,message=FALSE, warning=FALSE}
setwd('/Users/macbook/work/Code/Dataset2/')
set.seed(1234)
fam_all <- fread(paste0('gwa.qc.fam'))
N=dim(fam_all)[1];N
bim_all<- fread(paste0('gwa.qc.bim'))
M <- nrow(bim_all);M

indx.SNP=which((bim_all$V2)%in%c('rs1525873','rs2902663','rs343376','rs3813199'))

geno_mat<- readPlinkToMatrixByIndex('gwa.qc', 1:N, indx.SNP)

```

##### R functions to calculate the allele frequency.
```{r,message=FALSE, warning=FALSE}
# Compute allele frequencies for each SNP (column), ignoring NAs
allele_freq <- apply(geno_mat, 2, function(x) {
  sum(x, na.rm = TRUE) / (2 * sum(!is.na(x)))
})

# Look at the result
allele_freq

```


##### We can also conduct the HWE test in R.
```{r,message=FALSE, warning=FALSE}

hwe_pvalue <- function(geno_vector) {
  # Remove missing
  geno_vector <- geno_vector[!is.na(geno_vector)]
  # Count genotypes
  counts <- table(geno_vector)
  
  # Check if there are enough counts to test
  if (any(counts == 0)) {
    return(NA)  # Cannot test properly with empty cell
  } else {
    result=HWExact(as.numeric(counts))$pval
    #result=HWChisq(as.vector(counts))$pval
    return(result)
  }
}

geno_mat[geno_mat==-9]=NA
# Apply to all 4 SNPs (columns of the matrix)
hwe_pvalues <- apply(geno_mat, 2, hwe_pvalue)

# Show results
hwe_pvalues

#Compare with the HWE p-values produced by PLINK
hwe[hwe$ID%in%c('rs1525873','rs2902663','rs343376','rs3813199'),]
```


### 7. Logistic Regression: Additive Model

```{r,message=FALSE, warning=FALSE}
#Pick SNP rs1525873
table(geno_subset$rs1525873_T)
freq[freq$ID=='rs1525873',]
```
The T allele is the major allele, with frequency 1 - 0.247497. Therefore, most samples have genotype coded as 2 rather than 0.



```{r,message=FALSE, warning=FALSE}
# Fit logistic regression model
model <- glm(PHENOTYPE ~ rs1525873_T, data = geno_subset, family = binomial)

# View model summary
summary(model)
table(geno_subset$rs1525873_T)
```
##### We visualize the association between the phenotype and the genotype.
```{r,message=FALSE, warning=FALSE}
# 1️⃣ Make prediction over continuous range (for smooth curve)
 geno_range <- data.frame(
   rs1525873_T = seq(0, 2, length.out = 100)
 )
 geno_range$predicted_prob <- predict(model, newdata = geno_range, type = "response")


count_data <- geno_subset %>%
  mutate(PHENOTYPE = round(PHENOTYPE)) %>%  
  filter(!is.na(rs1525873_T), !is.na(PHENOTYPE)) %>%
  group_by(rs1525873_T, PHENOTYPE) %>%
  summarise(count = n(), .groups = "drop")

ggplot() +
  geom_point(
    data = count_data,
    aes(x = rs1525873_T, y = PHENOTYPE, size = count),
    color = "blue"
  ) +
  geom_line(
    data = geno_range,
    aes(x = rs1525873_T, y = predicted_prob),
    color = "red",
    size = 1.2
  ) +
  scale_size_continuous(range = c(2, 10)) +
  labs(
    title = "Logistic Regression: PHENOTYPE ~ Genotype",
    x = "Genotype (0/1/2)",
    y = "Phenotype ",
    size = "Count"
  ) +
  coord_cartesian(xlim = c(-0.5, 2.5), ylim = c(-0.1, 1.1)) +
  theme_minimal()
```

**Interpretation**: We performed logistic regression to test the association between genotype (rs1525873_T) and the binary phenotype. The regression coefficient for rs1525873_T is 0.14749 (p = 0.00476), indicating that each additional copy of the T allele is associated with increased odds of being a case. The significant p-value suggests evidence of an association between this SNP and the phenotype in our sample.

In practice, when testing genetic associations, we include additional covariates in the model—such as age, sex, and principal components to adjust for confounding effects. We will cover this in more detail later in the course.

### 8. Logistic Regression: Dominance and Recessive Models

##### Dominance coding 
```{r}
geno_subset_dom <- geno_subset  
geno_subset_dom[, 7:ncol(geno_subset_dom)] <- as.data.frame(
  lapply(geno_subset[, 7:ncol(geno_subset)], function(x) ifelse(x > 0, 1, 0))
)
head(geno_subset_dom)

```

Fit logistic regression model:

```{r}
model <- glm(PHENOTYPE ~ rs1525873_T, data = geno_subset_dom, family = binomial)
summary(model)
```
Visualize the assocaition between the phenotype and the genotype:
```{r}
# 1️⃣ Make prediction over continuous range (for smooth curve)
 geno_range <- data.frame(
   rs1525873_T = seq(0, 1, length.out = 100)
 )
 geno_range$predicted_prob <- predict(model, newdata = geno_range, type = "response")

count_data <- geno_subset_dom %>%
  mutate(PHENOTYPE = round(PHENOTYPE)) %>%   # 确保是0/1
  filter(!is.na(rs1525873_T), !is.na(PHENOTYPE)) %>%  # 移除NA
  group_by(rs1525873_T, PHENOTYPE) %>%
  summarise(count = n(), .groups = "drop")

print(count_data)

ggplot() +
  geom_point(
    data = count_data,
    aes(x = rs1525873_T, y = PHENOTYPE, size = count),
    color = "blue"
  ) +
  geom_line(
    data = geno_range,
    aes(x = rs1525873_T, y = predicted_prob),
    color = "red",
    size = 1.2
  ) +
  scale_size_continuous(range = c(2, 10)) +
  labs(
    title = "Logistic Regression: PHENOTYPE ~ Genotype",
    x = "Genotype (0/1)",
    y = "Phenotype ",
    size = "Count"
  ) +
  coord_cartesian(xlim = c(-0.5, 1.5), ylim = c(-0.1, 1.1)) +
  theme_minimal()
```



**Interpretation**: In the additive model, we see a clear dose-response relationship between genotype and disease risk: each additional copy of the risk allele increases the log-odds of being a case, as reflected in the significant and positive regression coefficient (~2.0). By contrast, the dominance model treats any presence of the risk allele (Aa or AA) the same, leading to a steeper estimated effect (~2.98). The AIC is higher under the dominance model (AIC=3663.2), indicating the additive model (AIC=3206.8) may fit slightly better overall by capturing a more gradual allele-dose effect.


##### Recessive Coding:
```{r}
geno_subset_dom <- geno_subset  # make a copy
geno_subset_dom[, 7:ncol(geno_subset_dom)] <- as.data.frame(
  lapply(geno_subset[, 7:ncol(geno_subset)], function(x) ifelse(x == 2, 1, 0))
)
```

```{r}
# Fit logistic regression model
model <- glm(PHENOTYPE ~ rs1525873_T, data = geno_subset_dom, family = binomial)
summary(model)

geno_range <- data.frame(
   rs1525873_T = seq(0, 1, length.out = 100)
 )
 geno_range$predicted_prob <- predict(model, newdata = geno_range, type = "response")


count_data <- geno_subset_dom %>%
  mutate(PHENOTYPE = round(PHENOTYPE)) %>%   # 确保是0/1
  filter(!is.na(rs1525873_T), !is.na(PHENOTYPE)) %>%  # 移除NA
  group_by(rs1525873_T, PHENOTYPE) %>%
  summarise(count = n(), .groups = "drop")

print(count_data)

ggplot() +
  geom_point(
    data = count_data,
    aes(x = rs1525873_T, y = PHENOTYPE, size = count),
    color = "blue"
  ) +
  geom_line(
    data = geno_range,
    aes(x = rs1525873_T, y = predicted_prob),
    color = "red",
    size = 1.2
  ) +
  scale_size_continuous(range = c(2, 10)) +
  labs(
    title = "Logistic Regression: PHENOTYPE ~ Genotype",
    x = "Genotype (0/1)",
    y = "Phenotype ",
    size = "Count"
  ) +
  coord_cartesian(xlim = c(-0.5, 1.5), ylim = c(-0.1, 1.1)) +
  theme_minimal()
```


**Interpretation**: In the recessive model, the predicted risk when genotype = 0 is higher than under the dominance model,when genotype = 0 includes both aa and Aa individuals. This means the average risk for genotype=0 in the recessive model includes some risk from heterozygotes (Aa), pushing the predicted probability higher than if it were only aa. The AIC of 3400.7 indicates that the additive model offers the best overall fit, consistent with the interpretation that heterozygotes (Aa) confer intermediate risk relative to aa.


In practice, the choice of genotype coding depends on biological hypotheses about the mode of inheritance and the statistical fit of different models. Researchers typically start with additive coding as the primary analysis, and may perform sensitivity analyses with dominance and recessive codings.